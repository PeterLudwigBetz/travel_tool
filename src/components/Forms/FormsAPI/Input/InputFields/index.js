import HtmlInput from './HtmlInput';
import DropdownSelect from './DropdownSelect';
import DateInput from './Date';
import ButtonToggler from './ButtonToggler';


export { DropdownSelect, DateInput, ButtonToggler, HtmlInput};
