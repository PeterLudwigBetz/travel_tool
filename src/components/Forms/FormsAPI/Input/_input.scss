form .form-input {
    display: flex;
    flex-flow: column;
    width: 100%;
    height: 64px;
    margin: 0 0 20px 0;
    transition: height .5s ease-out;
    &.full-width {
        width: 100%;
        height: 64px;
        transition: height .4s ease-out;
        &.other-dest {
            &--hidden,
            &--hiddenerror {
                height: 0px;
                overflow: hidden;
                margin: 0;
            }
            &--error {
                height: 84px;
            }
        }
    }
    span.error {
        color: red;
        font-size: 12px;
    }
    &.error {
        height: 84px;
        .input, input {
          border: 1px solid #FF5359;
        }
        .input:hover, input:hover {
          border: 1px solid #FF5359;
        }
        .input:focus, input:focus {
          box-shadow: 0px 0px 3.5px rgba(255,83,89,0.7);
          border: 1px solid #FF5359;
        }
    }
    input,
    .input {
        box-sizing: border-box;
        width: 100%;
        min-width: 260px;
        height: 36px;
        border-radius: 2px;
        outline: none;
        font-size: 14px;
        line-height: 18px;
        color: #4F4F4F;
        margin-top: 8px;
        font-family: "DIN Pro";
        font-weight: 300;
        cursor: pointer;
        &[type="text"] {
            cursor: text;
        }
        button {
            width: 130px;
            margin: 0;
            &.bg-btn {
                &--inactive {
                    background-color: #FFFFFF;
                    color: #4F4F4F;
                }
            }
        }
    }
    input:hover,
    .input.select:hover {
        border: 1px solid rgba(51, 89, 219, 0.7);
    }
    input:focus,
    .input.select:focus {
        box-shadow: 0px 0px 3.5px rgba(51, 89, 219, 0.6);
        border: 1px solid rgba(51, 89, 219, 0.7);
    }
    input,
    .input.select {
        border: 1px solid #D4D4D4;
        box-sizing: border-box;
        padding-left: 15px;
        position: relative;
        &::placeholder {
            color: #999999;
        }
        .value {
            position: relative;
            display: flex;
            align-items: center;
            height: 100%;
            .select-dropdown {
                position: absolute;
                right: 15px;
                transition: transform .05s ease-out;
                transform-origin: center;
            }
        }
    }
    label {
        span {
            color: #999999;
            font-family: "DIN Pro";
            font-size: 12px;
            line-height: 15px;
        }
    }
    .select-menu {
        z-index: 150;
        position: absolute;
        left: -1.05px;
        top: 26px;
        padding-left: 0px;
        box-sizing: border-box;
        display: flex;
        flex-flow: column;
        width: calc(100% + 3px);
        background-color: #FFFFFF;
        list-style-type: none;
        box-shadow: 0px 0px 1px 1px rgba(0, 0, 0, 0.1);
        border-radius: 2px;
        transition: transform .4s ease-out;
        transform-origin: top;
        box-sizing: border-box;
        li {
            display: flex;
            align-items: center;
            cursor: pointer;
            height: 32px;
            &:hover {
                background-color: #F8F8F8;
                color: #3359DB;
            }
            div {
                display: flex;
                align-items: center;
                padding-left: 15px;
                outline: none;
                width: 100%;
                height: 100%;
            }
        }
    }
    .date-wrapper {
        display: flex;
        position: relative;
        input {
            background: transparent;
            color: transparent;
            text-shadow: 0 0 0 #999999;
            cursor: context-menu;
        }
        .calendar-icon {
            position: absolute;
            right: 15px;
            top: 20px;
            z-index: -1;
        }
    }
    input:disabled {
      cursor: not-allowed;
      &:hover {
        border: 1px solid #D4D4D4;
      }
    }
    .input.select .select-menu {
        transition-delay: 0.25s;
        /* Don't reduce this */
    }
    .input.select .select-menu {
        transform: scaleY(0);
    }
    .input.select:focus .select-menu {
        transform: scaleY(1);
    }
    .react-datepicker-popper {
        box-shadow: .5px 0px 1px 0px rgba(0, 0, 0, 0.1);
        margin: 0;
        position: absolute;
    }
    .react-datepicker-popper[data-placement^="top"] {
        margin: 0;
        top: 5px !important;
    }
    .react-datepicker-popper[data-placement^="bottom"] {
        margin: 0;
        bottom: 1px !important;
    }
    .calendar-body {
        font-family: "DIN Pro";
        border: 1px solid #E4E4E4;
        border-radius: 2px;
        background-color: #FFFFFF;
        .calendar-day {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 36.5px;
            width: 35.5px;
            border-right: 1px solid #E4E4E4;
            border-radius: 0;
            box-sizing: border-box;
            background-color: #FFFFFF;
            margin: 0;
            font-size: 14px;
            color: #4F4F4F;
            &:first-child {
                border-left: 1px solid #E4E4E4;
            }
        }
        .react-datepicker__header {
            background-color: #FFFFFF;
            color: #3359DB;
            border: none;
        }
        .react-datepicker__month {
            margin: 0;
            padding: 5px;
        }
        .react-datepicker__triangle {
            display: none;
        }
        .react-datepicker__week {
            box-sizing: border-box;
            display: flex;
            width: 250px;
            border-top: 1px solid #E4E4E4;
            &:last-child {
                border-bottom: 1px solid #E4E4E4;
            }
        }
        .react-datepicker__day-names {
            display: flex;
            padding-left: 5px;
            .react-datepicker__day-name {
                display: flex;
                justify-content: center;
                width: 36.5px;
                height: 16px;
                margin: 0;
                color: #4F4F4F;
                font-family: "DIN Pro";
                font-size: 12px;
                font-weight: 300;
                line-height: 15px;
            }
        }
        .react-datepicker__day--disabled {
            cursor: default;
            color: #ccc !important;
        }
        .react-datepicker__current-month {
            display: flex;
            justify-content: center;
            align-items: center;
            height: 64px;
            background-color: #FFFFFF;
            color: #3359DB;
        }
        .react-datepicker__navigation {
            top: 32px;
            outline: none;
            &--next {
                border-left-color: #3359DB;
            }
            &--previous {
                border-right-color: #3359DB;
            }
        }
        .react-datepicker__day {
            box-sizing: border-box;
            &--selected {
                background-color: #F8F8F8;
                color: #3359DB;
            }
            &--outside-month {
                background-color: #FFFFFF;
                color: #E4E4E4;
            }
        }
    }
}

@media (min-width: 640px) {
    form .form-input {
        width: 260px;
    }
}

.required-field {
    margin-left: 61%;
    font-size: 9px;
    color: red;
    font-family: DIN Pro;
    @media( max-width: 479px) {
        margin-left: 40%
    }
    @media( max-width: 320px) {
        margin-left: 34%
    }
}
