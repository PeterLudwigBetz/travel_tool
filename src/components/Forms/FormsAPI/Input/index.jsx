import React, { PureComponent } from 'react';
import { PropTypes } from 'prop-types';
import { HtmlInput, DropdownSelect, DateInput, ButtonToggler } from './InputFields';
import createEventHandlersFor from '../formEventHandlers';
import './_input.scss';

class Input extends PureComponent {

  getInputType() {
    let { name, type, onChange } = this.props;
    const { errors, targetForm, validatorName } = this.context;

    // get event handlers for the form in this context, for its input named 'name'
    const eventHandlers = createEventHandlersFor(targetForm, name, validatorName);

    if(type === 'text'){
      this.props = {
        ...this.props,
        onBlur: eventHandlers.handleInputBlur,
        onChange: onChange ||  eventHandlers.handleInputChange
      };
      return HtmlInput;
    } else { // explore other input options
      return this.switchInputWithProps(type, eventHandlers);
    }
  }

  switchInputWithProps = (type, eventHandlers) => {
    const { onChange } = this.props;
    switch (type) {
    case 'button-toggler':
      // switches mutually exclusive options like a fancy set of radio buttons
      this.props = {
        ...this.props,
        onChange: onChange ||  eventHandlers.handleSelectTogglerOpt
      };
      return ButtonToggler;

    case 'date':
      this.props = { // save props for date type for use in render InputElement
        ...this.props,
        onChange: onChange || eventHandlers.handleSelectDate,
        onBlur: eventHandlers.handleInputBlur
      };
      return DateInput; // pick date Input

    case 'dropdown-select':
      this.props = {
        ...this.props,
        onChange: onChange ||  eventHandlers.handleSelectDropdown
      };
      return DropdownSelect;
    }
  }

  labelNote(note) { // a small message tied to an inputs label
    return note ? (
      <span>
        &nbsp;
        { note }
      </span>
    ) : null;
  }

  render() {
    const { errors } = this.context;
    let { name, label, labelNote, className } = this.props;
    let customClass = className ? className : '';
    // switch input types into InputElement
    const InputElement = this.getInputType();

    return (
      <div
        className={`form-input ${customClass} ${errors[name] ? 'error' : ''}`}
      >
        <label htmlFor={name}>
          {label}
          <span style={{color: 'red'}}>
          *
          </span>
          {this.labelNote(labelNote)}
        </label>
        <InputElement error={errors[name]} {...this.props} />
        <span className="error">
          {errors[name]}
        </span>
      </div>
    );
  }
}

Input.contextTypes = {
  errors: PropTypes.object.isRequired,
  targetForm: PropTypes.object.isRequired,
  validatorName: PropTypes.string
};

Input.propTypes = {
  name: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  onChange: PropTypes.oneOfType([
    PropTypes.func, PropTypes.object
  ]).isRequired,
  labelNote: PropTypes.string,
  className: PropTypes.string,
};

Input.defaultProps = {
  labelNote: '',
  className: ''
};

export default Input;
