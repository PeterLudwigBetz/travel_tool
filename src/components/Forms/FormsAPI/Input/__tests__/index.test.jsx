import React from 'react';
import Input from '..';

describe('<Input />', () => {
  let wrapper;

  beforeEach(() => {
    const contextObject = {
      context: {
        errors: {fullname: 'This field is required'},
        targetForm: {
          validate: ()=>{}
        }
      }
    };

    wrapper = mount(
      <Input
        name="fullname"
        type="text"
        onChange={()=>{}}
        label="Test label"
      />, contextObject
    );
  });


  it('renders as expected', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('Displays the error messge', () => {
    const errorText = wrapper.find('span.error').text();
    expect(errorText).toBe('This field is required');
  });

});
