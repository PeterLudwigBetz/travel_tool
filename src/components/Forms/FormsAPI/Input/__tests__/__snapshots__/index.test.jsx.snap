// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input /> renders as expected 1`] = `
<div
  className="form-input error"
>
  <label
    htmlFor="fullname"
  >
    Test label
  </label>
  <input
    className=""
    error=""
    label="Test label"
    name="fullname"
    onBlur={[Function]}
    onChange={[Function]}
    type="text"
  />
  <span
    className="error"
  >
    This field is required
  </span>
</div>
`;
